bridges/snowbridge/primitives/router/src/inbound/v2.rs

```diff
@@ -0,0 +1,257 @@
+	InvalidXCM,
+	/// The XCM provided with the message could not be decoded into versioned XCM.
+	InvalidVersionedXCM,
+	/// Invalid claimer MultiAddress provided in payload.
+	InvalidClaimer,
+	/// Invalid foreign ERC20 token ID
+	InvalidAsset,
+	/// The origin could not be added to the interior location of the origin location.
+	InvalidOrigin,
+}
+
+pub trait ConvertMessage {
+	fn convert(message: Message, origin_account: Location) -> Result<Xcm<()>, ConvertMessageError>;
+}
+
+pub struct MessageToXcm<
+	EthereumNetwork,
+	InboundQueuePalletInstance,
+	ConvertAssetId,
+	WethAddress,
+	GatewayProxyAddress,
+> where
+	EthereumNetwork: Get<NetworkId>,
+	InboundQueuePalletInstance: Get<u8>,
+	ConvertAssetId: MaybeEquivalence<TokenId, Location>,
+	WethAddress: Get<H160>,
+	GatewayProxyAddress: Get<H160>,
+{
+	_phantom: PhantomData<(
+		EthereumNetwork,
+		InboundQueuePalletInstance,
+		ConvertAssetId,
+		WethAddress,
+		GatewayProxyAddress,
+	)>,
+}
+
+impl<
+		EthereumNetwork,
+		InboundQueuePalletInstance,
+		ConvertAssetId,
+		WethAddress,
+		GatewayProxyAddress,
+	> ConvertMessage
+	for MessageToXcm<
+		EthereumNetwork,
+		InboundQueuePalletInstance,
+		ConvertAssetId,
+		WethAddress,
+		GatewayProxyAddress,
+	>
+where
+	EthereumNetwork: Get<NetworkId>,
+	InboundQueuePalletInstance: Get<u8>,
+	ConvertAssetId: MaybeEquivalence<TokenId, Location>,
+	WethAddress: Get<H160>,
+	GatewayProxyAddress: Get<H160>,
+{
+	fn convert(
+		message: Message,
+		origin_account_location: Location,
+	) -> Result<Xcm<()>, ConvertMessageError> {
+		let mut message_xcm: Xcm<()> = Xcm::new();
+		if message.xcm.len() > 0 {
+			// Decode xcm
+			let versioned_xcm = VersionedXcm::<()>::decode_with_depth_limit(
+				MAX_XCM_DECODE_DEPTH,
+				&mut message.xcm.as_ref(),
+			)
+			.map_err(|_| ConvertMessageError::InvalidVersionedXCM)?;
+			message_xcm = versioned_xcm.try_into().map_err(|_| ConvertMessageError::InvalidXCM)?;
+		}
+
+		log::debug!(target: LOG_TARGET,"xcm decoded as {:?}", message_xcm);
+
+		let network = EthereumNetwork::get();
+
+		// use weth as asset
+		let fee_asset = Location::new(
+			2,
+			[
+				GlobalConsensus(EthereumNetwork::get()),
+				AccountKey20 { network: None, key: WethAddress::get().into() },
+			],
+		);
+		let fee: XcmAsset = (fee_asset.clone(), message.fee).into();
+		let mut instructions = vec![
+			DescendOrigin(PalletInstance(InboundQueuePalletInstance::get()).into()),
+			UniversalOrigin(GlobalConsensus(network)),
+			ReserveAssetDeposited(fee.clone().into()),
+			PayFees { asset: fee },
+		];
+
+		for asset in &message.assets {
+			match asset {
+				Asset::NativeTokenERC20 { token_id, value } => {
+					let token_location: Location = Location::new(
+						2,
+						[
+							GlobalConsensus(EthereumNetwork::get()),
+							AccountKey20 { network: None, key: (*token_id).into() },
+						],
+					);
+					instructions.push(ReserveAssetDeposited((token_location, *value).into()));
```

Comment:
`ReserveAssetDeposited` takes multiple Assets as an argument. Instead of using multiple `ReserveAssetDeposited` with a single asset in each, do one `ReserveAssetDeposited` with multiple assets.